lib/validator/v3/validator.go

```diff
@@ -15,46 +15,50 @@
 package v3
 
 import (
+	"fmt"
 	"net"
 	"reflect"
 	"regexp"
 	"strings"
 
+	log "github.com/sirupsen/logrus"
+	"gopkg.in/go-playground/validator.v8"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8svalidation "k8s.io/apimachinery/pkg/util/validation"
+
 	api "github.com/projectcalico/libcalico-go/lib/apis/v3"
 	"github.com/projectcalico/libcalico-go/lib/errors"
 	cnet "github.com/projectcalico/libcalico-go/lib/net"
 	"github.com/projectcalico/libcalico-go/lib/numorstring"
 	"github.com/projectcalico/libcalico-go/lib/selector"
-	log "github.com/sirupsen/logrus"
-
-	"github.com/projectcalico/libcalico-go/lib/backend/model"
-	"gopkg.in/go-playground/validator.v8"
-	"k8s.io/apimachinery/pkg/apis/meta/v1"
-	cerrors "github.com/projectcalico/libcalico-go/lib/errors"
 )
 
-var validate *validator.Validate
+// Split the validation into three groups.  Since a struct validator will trump
+// any field validation or sub-structure validation we need to split these out.
+var validatorPrimary *validator.Validate
+var validatorSecondary *validator.Validate
+var validatorMetadata *validator.Validate
+
+// Maximum size of annotations.
+const totalAnnotationSizeLimitB int64 = 256 * (1 << 10) // 256 kB`
 
 var (
-	nameLabelFmt          = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
-	nameSubdomainFmt      = nameLabelFmt + "(\\." + nameLabelFmt + ")*"	// Rob's "normal" name
-	knpPolicyPrefix       = "knp\\.default\\."
-
-	// reference a "profile"; standard name regex (nameDotsRegex)
-	policyNameRegex       = regexp.MustCompile("(^"+nameLabelFmt+"$)|(^"+knpPolicyPrefix+nameLabelFmt+"$)")
-	nameNoDotsRegex       = regexp.MustCompile("^"+nameLabelFmt+"$")
-	nameDotsRegex         = regexp.MustCompile("^"+nameSubdomainFmt+"$")	// "validate: name"
-
-	labelValueRegex       = nameDotsRegex
-	labelRegex            = regexp.MustCompile("^[a-zA-Z0-9_./-]{1,512}$")
-	nameRegex             = regexp.MustCompile("^[a-zA-Z0-9_\\.-]{1,128}$")	// replace this with nameDotsRegex
-	labelsToApplyValRegex = regexp.MustCompile("^[a-zA-Z0-9_\\.-]{0,128}$")
-	namespacedNameRegex   = regexp.MustCompile(`^[a-zA-Z0-9_\\./-]{1,128}$`)
+	nameLabelFmt     = "[a-z0-9]([-a-z0-9]*[a-z0-9])?"
+	nameSubdomainFmt = nameLabelFmt + "(\\." + nameLabelFmt + ")*"
```

Comment:
consider renaming since `Subdomain` is a legacy from dns. but we can do this post-beta if this is the only thing.